### PR TITLE
Stop error fix

### PR DESCRIPTION
**Pull request by asgothian**
Fixes Issue #379 by encapsulating calls in try/catch.

### CODE REVIEW

File: .remote-sync.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "logger": {
+    "title": "Remote Sync"
+  },
+  "uploadOnSave": false,
+  "useAtomicWrites": false,
+  "deleteLocal": false,
+  "hostname": "ioTinker.lan",
+  "port": "22",
+  "target": "/opt/iobroker/node_modules/zigbee-herdsman",
```
Comment (Koenkk) [inline .remote-sync.json:10]: Please .gitignore this file.

File: package.json [hunk 2]
```diff
@@ -73,7 +73,7 @@
   "name": "zigbee-herdsman",
   "repository": {
     "type": "git",
-    "url": "git+https://github.com/Koenkk/zigbee-herdsman.git"
+    "url": "git+https://github.com/Asgothian/zigbee-herdsman.git"
```
Comment (Koenkk) [inline package.json:76]: Please revert this change

File: src/controller/controller.ts [hunk 3]
```diff
@@ -281,7 +283,9 @@ class Controller extends events.EventEmitter {
 
         clearInterval(this.backupTimer);
         clearInterval(this.databaseSaveTimer);
-        await this.backup();
+        try {
+            await this.backup();
+        } catch (e) {}
         await this.adapter.stop();
```
Comment (Koenkk) [inline src/controller/controller.ts:289]: In what situation does the backup fail but the adapter stop doesn't? I would expect the adapter.stop to also be in the try catch.

File: src/controller/model/endpoint.ts [hunk 4]
```diff
@@ -647,7 +647,7 @@ class Endpoint extends Entity {
         } catch (error) {
             error.message = `${log} failed (${error.message})`;
             debug.error(error.message);
-            throw error;
```
Comment (Koenkk) [inline src/controller/model/endpoint.ts:650]: This should not be removed.

File: src/controller/controller.ts [hunk 5]
```diff
@@ -265,7 +265,9 @@ class Controller extends events.EventEmitter {
 
     public async stop(): Promise<void> {
         this.stopping = true;
-        this.databaseSave();
+        try {
```
Comment (Koenkk) [inline src/controller/controller.ts:268]: Is this one really needed? (I don't see how this can fail unless e.g. wrong permission are setup). 

As an alternative we can move it below the adapter.stop().

File: package.json (hunk 2 above)
Comment (asgothian) [reply package.json:76]: Sorry, this change slipped through, will be changed back.

File: src/controller/controller.ts (hunk 5 above)
Comment (asgothian) [reply src/controller/controller.ts:268]: This is **the** key item of this PR.
If the start of the adapter fails before the network is initialised, but after the serial port was opened, some internal structures are not initialised leading to a number of errors. The exact error depends on how far the initialisation has come and possibly also on timing. I was not able to pinpoint the exact placement, which is why i went for the 'catch all' approach.

```
TypeError: Cannot read property 'getEntries' of null
    at Function.loadFromDatabaseIfNecessary (/opt/iobroker/node_modules/zigbee-herdsman/dist/controller/model/device.js:220:55)
```

Moving it behind the adapter.stop() might be possible, but would only solve half the issue. The Herdsman would still fail due to its internal structures not being initialised in a situation where this should not pose an issue to begin with.

File: src/controller/controller.ts (hunk 3 above)
Comment (asgothian) [reply src/controller/controller.ts:289]: Interestingly enough, the adapter.stop() does not throw an exception in any of the cases i was able to duplicate, while the this.backup still references uninitialised data structures. I could imagine placing it in the try/catch, but am hesitant to do so as i would want an error thrown on adapter.stop to be passed to the client application, while the error on backup would only be valid for cases where the backup should succeed.

Since the error messages are not currently parseable, this becomes impossible to determine, so i opted for catching the errors from being unable to perform a backup while keeping the errors on stop for the client application to handle.

File: src/controller/model/endpoint.ts (hunk 4 above)
Comment (asgothian) [reply src/controller/model/endpoint.ts]: I was not aware this was still in the code. It is a remnant of an older test and definitely not wanted.

File: src/controller/controller.ts (hunk 5 above)
Comment (Koenkk) [reply src/controller/controller.ts:268]: Can you try without this try/catch and changing https://github.com/Koenkk/zigbee-herdsman/blob/941613e8f75047f529a8f0c87d57192508c9b78d/src/controller/controller.ts#L114 to 


```js
        // Database
        this.database = Database.open(this.options.databasePath);
        Entity.injectDatabase(this.database);

        // Adapter
        this.adapter = await Adapter.create(this.options.network,
            this.options.serialPort, this.options.backupPath, this.options.adapter, this.logger);
        debug.log(`Starting with options '${JSON.stringify(this.options)}'`);
        const startResult = await this.adapter.start();
        debug.log(`Started with result '${startResult}'`);
        Entity.injectAdapter(this.adapter);
```

Comment (asgothian) [reply src/controller/controller.ts:268]: I'll check this later today.

File: src/controller/controller.ts (hunk 3 above)
Comment (asgothian) [reply src/controller/controller.ts:289]: I managed to duplicate the backup failure tonight too. The failure originated in the test if use of backup was supported. On the zStack Adapter (the only one where i managed to recreate it), a crash was caused here:
```
    public async supportsBackup(): Promise<boolean> {
        return this.version.product !== ZnpVersion.zStack12;
    }
```
by `this.version` being `undefined`.

Due to other possible side effects i decided to keep this try/catch in the PR, but added an error message to be posted for information.

File: src/controller/controller.ts (hunk 5 above)
Comment (asgothian) [reply src/controller/controller.ts:268]: tested - works.

File: src/controller/controller.ts (hunk 3 above)
Comment (Koenkk) [reply src/controller/controller.ts:292]: What if you change `return this.version.product !== ZnpVersion.zStack12;` to `return this.version && this.version.product !== ZnpVersion.zStack12;`

Comment (asgothian) [reply src/controller/controller.ts:292]: It will remove this one error, but i am hesitant to go that route, as it is just one of many errors i have seen which i am actually able to recreate. I am not convinced that i will not hit yet another error when this.version is populated but deeper in the backup functionality something is still uninitialized.

Comment (Koenkk) [reply src/controller/controller.ts:292]: @asgothian then we can fix that later, I think it's better than putting a try/catch for now.

Comment (asgothian) [reply src/controller/controller.ts:292]: Ok.. i will run a test tomorrow - awaiting guests for tonight

Comment (asgothian) [reply src/controller/controller.ts]: Updated. Was unable to test, as i could not get the error to reoccur. But i am sufficiently certain that it will work.

**Comment by Koenkk** [comment]
Thanks!
